### PR TITLE
Add docker based package builder and switch to system compiler

### PR DESCRIPTION
**Pull request by xinyazhang**
## Major Changes:

* `dockerfile/build.sh` script that builds AOTriton from offical AlmaLinux 8 docker image.
* Deprecate hipcc since GCC >= 9 / Clang >= 10 is capable to compile the C++ part
    + `hipcc` is not compatible with SCL. `scl enable gcc-toolset-13 "/opt/rocm/bin/hipcc -v"` is supposed to show "Found candidate GCC installation" but nothing is displayed, and eventually this triggers https://github.com/ROCm/ROCm/issues/1843 during compiling.

## Minor Changes

* In `pyaotriton`, register `hipError_t` locally to avoid "`hipError_t` is already registered" bug.

### CODE REVIEW

File: dockerfile/manylinux_2_28.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+RUN dnf install -y gcc-toolset-13 python3.11 python3.11-devel \
+    zstd libzstd-devel xz-devel zlib-devel git which vim && \
+    update-alternatives --set python /usr/bin/python3.11 && \
+    update-alternatives --set python3 /usr/bin/python3.11 && \
+    python -m ensurepip && \
+    python -m pip install cmake ninja wheel
+RUN mkdir /root/build
+
+ARG amdgpu_installer="https://repo.radeon.com/amdgpu-install/6.2.2/el/8.10/amdgpu-install-6.2.60202-1.el8.noarch.rpm"
```
Comment (jithunnair-amd) [inline dockerfile/manylinux_2_28.Dockerfile:11]: This is a maintenance headache, we need to make this more generic for this to be generally usable for any ROCm release

Comment (jithunnair-amd) [reply dockerfile/manylinux_2_28.Dockerfile:11]: Discussed this offline, this is the default value. The intention is for any CI job or users to provide a link to the desired amdgpu installer for their desired ROCm version.
We could use logic similar to this file:
https://raw.githubusercontent.com/wiki/ROCmSoftwarePlatform/pytorch/files/install_kdb_files_for_pytorch_wheels.sh
to keep the default argument always up-to-date.

File: dockerfile/input/install_aotriton.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  if [ -f "$f" ]; then
+    echo "apply patch $f"
+    git apply "$f"
+  fi
+done
+
+cd /root/build
+scl enable gcc-toolset-13 "cd aotriton/third_party/triton/python; python setup.py develop --user"
+scl enable gcc-toolset-13 "mkdir -p aotriton/build; cd aotriton/build; cmake .. -DCMAKE_PREFIX_PATH=/opt/rocm -DPYTHON_EXECUTABLE=/usr/bin/python3.11 -DCMAKE_INSTALL_PREFIX=installed_dir/aotriton -DCMAKE_BUILD_TYPE=Release -DAOTRITON_GPU_BUILD_TIMEOUT=0 \"-DTARGET_GPUS=${AOTRITON_TARGET_GPUS}\" -DAOTRITON_NO_PYTHON=ON -G Ninja && ninja install"
+rocmver=$(scl enable gcc-toolset-13 "cpp -I/opt/rocm/include /input/print_rocm_version.h"|tail -n 1|sed 's/ //g')
```
Comment (jithunnair-amd) [inline dockerfile/input/install_aotriton.sh:20]: @naromero77amd This is another interesting and even more compact way of getting the ROCM_VERSION* defines. Something to consider if we want to make our LoadHIP.cmake logic even more compact.

File: v2src/CMakeLists.txt [hunk 3]
```diff
@@ -168,13 +168,14 @@ target_include_directories(aotriton_v2 PUBLIC ${CMAKE_CURRENT_LIST_DIR}/../inclu
...
-target_compile_options(aotriton_v2 PRIVATE -fPIC --no-offload-arch=all)
+target_compile_options(aotriton_v2 PRIVATE -fPIC)
 if(AOTRITON_BUILD_FOR_TUNING)
   target_compile_definitions(aotriton_v2 PRIVATE -DAOTRITON_BUILD_FOR_TUNING=1)
 else(AOTRITON_BUILD_FOR_TUNING)
   target_compile_definitions(aotriton_v2 PRIVATE -DAOTRITON_BUILD_FOR_TUNING=0)
 endif(AOTRITON_BUILD_FOR_TUNING)
-target_link_libraries(aotriton_v2 lzma_interface)
+target_link_libraries(aotriton_v2 PRIVATE lzma_interface)
+target_link_libraries(aotriton_v2 PUBLIC hip::host hip::amdhip64)
```
Comment (jithunnair-amd) [inline v2src/CMakeLists.txt:178]: We need to introduce an explicit dependency on the HIP targets because we are not using the hipcc compiler as the default compiler anymore: https://github.com/ROCm/aotriton/pull/51/files#diff-3c713483b9080c3c648c0ac4fb576cc9fbe0ad2b2b4f661637939993fbbc6fbcR19

File: CMakeLists.txt [hunk 4]
```diff
@@ -31,9 +33,6 @@ if(NOT AOTRITON_NO_PYTHON)
   add_subdirectory(third_party/pybind11)
 endif()
 
-# Must be after pybind11
-set(CMAKE_CXX_COMPILER hipcc)
```
Comment (jithunnair-amd) [inline CMakeLists.txt:35]: We do not need to use `hipcc` since we are not compiling any kernels, so a host compiler should be good enough.

File: v2src/CMakeLists.txt [hunk 5]
```diff
@@ -168,13 +168,14 @@ target_include_directories(aotriton_v2 PUBLIC ${CMAKE_CURRENT_LIST_DIR}/../inclu
 target_include_directories(aotriton_v2 PUBLIC ${CMAKE_BINARY_DIR}/include)  # for <build>/include/aotriton/config.h. Code should use <aotriton/config.h>
 target_include_directories(aotriton_v2 PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
 target_include_directories(aotriton_v2 PRIVATE ${CMAKE_CURRENT_LIST_DIR}/../third_party/incbin)
-target_compile_options(aotriton_v2 PRIVATE -fPIC --no-offload-arch=all)
```
Comment (jithunnair-amd) [inline v2src/CMakeLists.txt:171]: Removed since we're not using hipcc anymore

File: v2src/flash/attn_bwd.cc [hunk 6]
```diff
@@ -156,12 +156,12 @@ bwd_kernel_dk_dv(T4 q,
     .D = &delta,
     .num_head_q = num_head_q,
     .num_head_k = num_head_k,
+    .head_dim = head_size,
     .cu_seqlens_q = &cu_seqlens_q,
     .cu_seqlens_k = &cu_seqlens_k,
     .num_seqlens = num_seqlens,
     .max_seqlen_q = max_seqlen_q,
     .max_seqlen_k = max_seqlen_k,
-    .head_dim = head_size,
```
Comment (jithunnair-amd) [inline v2src/flash/attn_bwd.cc:164]: Required because we're switching to gcc, which is more strict about the order of struct fields

File: bindings/hipruntime.cc [hunk 7]
```diff
@@ -8,7 +8,7 @@ namespace py = pybind11;
 namespace pyaotriton {
 #define EV(name) value(#name, name)
   void def_hipruntime(py::module_& m) {
-    py::enum_<hipError_t>(m, "hipError_t")
+    py::enum_<hipError_t>(m, "hipError_t", py::module_local())
```
Comment (jithunnair-amd) [inline bindings/hipruntime.cc:11]: If this is *not* due to the compiler switch or any other changes in this PR, then it would be better to separate this into its own PR with reasoning.

**Review by jithunnair-amd — APPROVED** [review]

**Comment by xinyazhang** [comment]
Dockerfile tested locallly and can build 0.7.1b
Compiler switch tested locally and can run test `pytest ../test/test_backward.py -k 1.2 -v -x`

**Comment by jithunnair-amd** [comment]
@ethanwee1 Can you please validate this PR by trying to build aotriton 0.7b and 0.7.1b using the Dockerfile in this PR? Please make sure that it generates a tarball with the following files:
```
$ tar tvf output/aotriton-0.7.1b-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz 
drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/
drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/lib/
-rw-r--r-- root/root 382452768 2024-10-29 09:11 aotriton/lib/libaotriton_v2.so
drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/
drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/aotriton/
drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/aotriton/_internal/
-rw-r--r-- root/root      1490 2024-10-29 08:48 aotriton/include/aotriton/_internal/triton_kernel.h
-rw-r--r-- root/root       592 2024-10-29 08:48 aotriton/include/aotriton/_internal/util.h
-rw-r--r-- root/root       566 2024-10-29 08:48 aotriton/include/aotriton/cpp_tune.h
-rw-r--r-- root/root       422 2024-10-29 08:48 aotriton/include/aotriton/dtypes.h
-rw-r--r-- root/root      5435 2024-10-29 08:48 aotriton/include/aotriton/flash.h
-rw-r--r-- root/root       695 2024-10-29 08:48 aotriton/include/aotriton/runtime.h
-rw-r--r-- root/root      3316 2024-10-29 08:48 aotriton/include/aotriton/util.h
```

**Comment by ethanwee1** [comment]
> @ethanwee1 Can you please validate this PR by trying to build aotriton 0.7b and 0.7.1b using the Dockerfile in this PR? Please make sure that it generates a tarball with the following files:
> 
> ```
> $ tar tvf output/aotriton-0.7.1b-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz 
> drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/
> drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/lib/
> -rw-r--r-- root/root 382452768 2024-10-29 09:11 aotriton/lib/libaotriton_v2.so
> drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/
> drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/aotriton/
> drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/aotriton/_internal/
> -rw-r--r-- root/root      1490 2024-10-29 08:48 aotriton/include/aotriton/_internal/triton_kernel.h
> -rw-r--r-- root/root       592 2024-10-29 08:48 aotriton/include/aotriton/_internal/util.h
> -rw-r--r-- root/root       566 2024-10-29 08:48 aotriton/include/aotriton/cpp_tune.h
> -rw-r--r-- root/root       422 2024-10-29 08:48 aotriton/include/aotriton/dtypes.h
> -rw-r--r-- root/root      5435 2024-10-29 08:48 aotriton/include/aotriton/flash.h
> -rw-r--r-- root/root       695 2024-10-29 08:48 aotriton/include/aotriton/runtime.h
> -rw-r--r-- root/root      3316 2024-10-29 08:48 aotriton/include/aotriton/util.h
> ```

Validated 0.7.1b after following build.sh 
`aotriton-0.7.1b-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz
`

**Comment by jithunnair-amd** [comment]
> > @ethanwee1 Can you please validate this PR by trying to build aotriton 0.7b and 0.7.1b using the Dockerfile in this PR? Please make sure that it generates a tarball with the following files:
> > ```
> > $ tar tvf output/aotriton-0.7.1b-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz 
> > drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/
> > drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/lib/
> > -rw-r--r-- root/root 382452768 2024-10-29 09:11 aotriton/lib/libaotriton_v2.so
> > drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/
> > drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/aotriton/
> > drwxr-xr-x root/root         0 2024-10-29 09:11 aotriton/include/aotriton/_internal/
> > -rw-r--r-- root/root      1490 2024-10-29 08:48 aotriton/include/aotriton/_internal/triton_kernel.h
> > -rw-r--r-- root/root       592 2024-10-29 08:48 aotriton/include/aotriton/_internal/util.h
> > -rw-r--r-- root/root       566 2024-10-29 08:48 aotriton/include/aotriton/cpp_tune.h
> > -rw-r--r-- root/root       422 2024-10-29 08:48 aotriton/include/aotriton/dtypes.h
> > -rw-r--r-- root/root      5435 2024-10-29 08:48 aotriton/include/aotriton/flash.h
> > -rw-r--r-- root/root       695 2024-10-29 08:48 aotriton/include/aotriton/runtime.h
> > -rw-r--r-- root/root      3316 2024-10-29 08:48 aotriton/include/aotriton/util.h
> > ```
> 
> Validated 0.7.1b and 0.7b after following build.sh
> 
> ```
> aotriton-0.7.1b-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz 
> aotriton-0.7b-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz
> ```

@xinyazhang Taking this out of Draft mode as it is ready to merge (even if queued) based on Ethan's validation

**Comment by xinyazhang** [comment]
@jithunnair-amd nope, queued PR should not be taken out of draft because it's based on previous work (for this PR specifically its https://github.com/ROCm/aotriton/pull/50)

**Comment by ethanwee1** [comment]
Tested with these commands to build 
`aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz`

Commands:

```
git clone https://github.com/ROCm/aotriton.git
cd aotriton/
git checkout xinyazhang/manylinux_2_28-dockerfile
cd dockerfile/
export AMDGPU_INSTALLER=https://repo.radeon.com/amdgpu-install/6.2.4/el/8.10/amdgpu-install-6.2.60204-1.el8.noarch.rpm
TRITON_LLVM_HASH="b5cc222d" bash build.sh input tmpfs output e278d4a853170c7a9063cfe847419414cb7b62b6 "MI300X;MI200" 2>&1 | tee buildlog2.log
tar tvf output/*.tar*
```

Output:
[aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared.txt](https://github.com/user-attachments/files/17738213/aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared.txt)
[buildlog2.log](https://github.com/user-attachments/files/17738214/buildlog2.log)

**Comment by jithunnair-amd** [comment]
> Tested with these commands to build `aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz`
> 
> Commands:
> 
> ```
> git clone https://github.com/ROCm/aotriton.git
> cd aotriton/
> git checkout xinyazhang/manylinux_2_28-dockerfile
> cd dockerfile/
> export AMDGPU_INSTALLER=https://repo.radeon.com/amdgpu-install/6.2.4/el/8.10/amdgpu-install-6.2.60204-1.el8.noarch.rpm
> mkdir -p output
> TRITON_LLVM_HASH="b5cc222d" bash build.sh input tmpfs output e278d4a853170c7a9063cfe847419414cb7b62b6 "MI300X;MI200" 2>&1 | tee buildlog2.log
> tar tvf output/*.tar*
> ```
> 
> Output: [aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared.txt](https://github.com/user-attachments/files/17738213/aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared.txt) [buildlog2.log](https://github.com/user-attachments/files/17738214/buildlog2.log)

Some notable lines in the log:
* `-rw-r--r-- root/root  13064056 2024-11-13 18:01 aotriton/lib/libaotriton_v2.so`: libaotriton_v2.so reduced to 13MB

**Comment by xinyazhang** [comment]
13 MiB is also significant considering it's functionality. I believe most of the size comes from the generated dispatching code.
I've added this into the [keepbook](https://github.com/ROCm/frameworks-internal/issues/7693#issuecomment-2442422526), but no concrete plan to implement it.

**Comment by ethanwee1** [comment]
Tested with these commands to build 
`aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared.tar.gz`

Commands:

```
git clone https://github.com/ROCm/aotriton.git
cd aotriton/
git checkout xinyazhang/manylinux_2_28-dockerfile
cd dockerfile/
export AMDGPU_INSTALLER=https://repo.radeon.com/amdgpu-install/6.2.4/el/8.10/amdgpu-install-6.2.60204-1.el8.noarch.rpm
mkdir -p output
TRITON_LLVM_HASH="b5cc222d" AOTRITON_NOIMAGE_MODE=ON bash build.sh input tmpfs output e278d4a853170c7a9063cfe847419414cb7b62b6 "MI300X;MI200" 2>&1 | tee buildlog2.log
tar tvf output/*.tar*
```

Output:
Size: 107MB
[aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared_WITHAOTRITON_NOIMAGE_MODE_ON.txt](https://github.com/user-attachments/files/17740386/aotriton-e278d4a853170c7a9063cfe847419414cb7b62b6-manylinux_2_28_x86_64-rocm6.2-shared_WITHAOTRITON_NOIMAGE_MODE_ON.txt)
[buildlog2.log](https://github.com/user-attachments/files/17740387/buildlog2.log)
